### PR TITLE
Reduce recovery time with compress or secure transport

### PR DESCRIPTION
**Pull request by dnhatn**
Today file-chunks are sent sequentially one by one in peer-recovery. This is a correct choice since the implementation is straightforward and recovery is network bound in most of the time. However, if the transport communication is secure, we might not be able to saturate the network bandwidth because encrypting/decrypting are compute-intensive.

With this commit, a source node can send multiple (default to 2) file-chunks without waiting for the acknowledgments from the target.

Below are the benchmark results for [PMC](https://github.com/elastic/rally-tracks/tree/master/pmc) and [NYC_taxis](https://github.com/elastic/rally-tracks/tree/master/nyc_taxis) datasets. The benchmark consists of two GCP instances (8CPU, 32GB RAM, 12GiB bandwidth and local SSD).

- PMC (20.2 GB)

| Transport | Baseline | chunks=1 | chunks=2 | chunks=3 | chunks=4 | 
| ----------| ---------| ---------- | --------- | ----------- | --------- |
| Plain         | 184s | 137s | 106s | 105s | 106s | 
| TLS           | 346s| 294s| 176s | 153s | 117s| 
| Compress| 1556s | 1407s | 1193s | 1183s | 1211s |
| Compress + TLS | n/a | n/a | n/a|n/a|n/a|

- NYC_Taxis (38.6GB)

| Transport | Baseline | chunks=1 | chunks=2 | chunks=3 | chunks=4 | 
| ----------| ---------| ---------- | --------- | ----------- | --------- |
| Plain         | 321s | 249s | 191s | * | * | 
| TLS           | 618s| 539s| 323s | 290s | 213s| 
| Compress| 2622s | 2421s | 2018s | 2029s | n/a |
| Compress + TLS | n/a | n/a | n/a|n/a|n/a|

Relates #33844

### CODE REVIEW

**Comment by dnhatn** [comment]
run gradle build tests 2

**Review by original-brownbear — APPROVED** [review]
LGTM, just some minor suggestions
(probably still best to wait for a second opinion from someone else :))

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: perhaps an `AtomicReference<Exception>` and then further down below just do `error.compareAndSet(null, (Exception) ex);`

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: CompletableFuture has problems with swallowing Errors (see https://github.com/elastic/elasticsearch/pull/32512). Let's use `ListenableFuture` for now.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: do we need the "clearing" part? I think there's no need to set this back to null

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: should we clean this up in `closeInternal`?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: assert that we just removed ourselves here?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: If a message is dropped or delayed for quite a bit, we possibly keep on adding and adding to this queue, which can become quite big. I wonder if we need more control on the sender side to prevent sending chunks until some prior chunks have been handled.

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks @dnhatn. I like the simplicity of this PR. I've left some comments and would like to ask you to also run tests with `tcp.compress` enabled (both for TLS enabled/disabled). It would also be good to see how much throughput we get at these numbers. For example, can we saturate a 10Gbit network with `num_chunks = 2`?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: +1 Good call. I will work on this.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java]: +1

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: done

**Comment by dnhatn** [comment]
@ywelsch I have addressed all your comments. Would you please have another look?

**Comment by dnhatn** [comment]
I ran a more realistic benchmark which consists of two GCP instances (8CPU, 32GB RAM, 10GiB bandwidth and local SSD).  Below are the results for PMC and NYC_taxis datasets.

- PMC (20.2 GB)

| Transport | Baseline | chunks=1 | chunks=2 | chunks=3 | chunks=4 | 
| ----------| ---------| ---------- | --------- | ----------- | --------- |
| Plain         | 184s | 137s | 106s | 105s | 106s | 
| TLS           | 346s| 294s| 176s | 153s | 117s| 
| Compress| 1556s | 1407s | 1193s | 1183s | 1211s |
| Compress + TLS | n/a | n/a | n/a|n/a|n/a|

- NYC_Taxis (38.6GB)

| Transport | Baseline | chunks=1 | chunks=2 | chunks=3 | chunks=4 | 
| ----------| ---------| ---------- | --------- | ----------- | --------- |
| Plain         | 321s | 249s | 191s | * | * | 
| TLS           | 618s| 539s| 323s | 290s | 213s| 
| Compress| n/a | 2421s | 2018s | 2029s | n/a |
| Compress + TLS | n/a | n/a | n/a|n/a|n/a|

The current approach does not reduce the recovery time much with compression because the compression is too expensive. It takes [20 minutes](https://gist.github.com/dnhatn/add58605b25afcbf23c63a41d9b5f3f7) to compress 20GB of Lucene index and compressing happens on (and blocks) the recovery thread. The recovery time is reduced linearly with the `max_concurrent_file_chunks` if we compress the file chunk requests in parallel. However, I don't think we should do it as it would take all the CPU which we should reserve for higher priority tasks such as search.  Then should we disable compression for the file-chunk requests regardless of the compression setting given that it's too expensive and only saves around [16\%](https://gist.github.com/dnhatn/add58605b25afcbf23c63a41d9b5f3f7) of bandwidth? Notes that we compress chunk by chunk, not the whole file. Here is the benchmark that sends file-chunk requests in parallel for PMC.

| Transport | Baseline |async chunks=2 | async chunks=3 | async chunks=4 |
| ----------| ---------|  --------- | ----------- | --------- |
| Compress| 1556s | 737s | 514s | 207s |

**Comment by dnhatn** [comment]
> Is the index using best_compression? Is it force-merged?

No, the index uses the default index_codec and no force_merged is called.

**Comment by ywelsch** [comment]
With the recent change, it does not pipeline sending requests for different files (i.e. one file needs to be completed before we start with next one)?

**Comment by dnhatn** [comment]
> i.e. one file needs to be completed before we start with next one

Yes, the previous change and this change both wait for the completion of the current file before sending the next file. This is because we wait for all outstanding requests when we close the current RecoveryOutputStream.

**Comment by ywelsch** [comment]
How difficult would it be to lift that limitation?

**Comment by dnhatn** [comment]
> How difficult would it be to lift that limitation?

@ywelsch I reworked this PR using the LocalCheckpointTracker as we discussed. Could you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: We should avoid accessing Lucene under cancellablethreads, see https://issues.apache.org/jira/browse/LUCENE-2239.

Just wrap the calls to `requestSeqIdTracker.waitForOpsToComplete` instead

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: why is this still needed? Just use `requestSeqIdTracker` (see my other comments)

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: just `requestSeqIdTracker.waitForOpsToComplete(requestSeqIdTracker.getMaxSeqNo());`

Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I wonder if there are issues here with waiting for everything to complete when there is cancellation / exceptions going on. We don't particularly care about pending outgoing requests when we terminate exceptionally, as when those return we won't do anything with them anyway.

I think we should execute this here only on the happy path

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: why change this? Let's throw the original exception like it used to be before

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left more comments but like the new approach.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: ++

**Comment by dnhatn** [comment]
@ywelsch I have removed the semaphore and simplified the flow a bit. Would you please give it another go?

File: docs/reference/modules/indices/recovery.asciidoc
Comment (ywelsch) [inline docs/reference/modules/indices/recovery.asciidoc]: also add a description of what this setting does? :)

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: do we need to close this one?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: catch Exception here? (We did it like that before)

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: add a comment here saying why you don't want to have this in an `else` branch

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: let's also put this call under `cancellableThreads.executeIO(()`
I know we don't block on it anymore, but let's be extra safe.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: put this at the beginning of the finally block?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Left some minor comments, getting close. Can you also check that the rate limiter is still working correctly by doing an explicit test with rate limiting and max_chunks set to something high (i.e. 5)

**Comment by dnhatn** [comment]
@ywelsch 
I re-ran the recovery benchmark. The new implementation is slightly (but insignificantly) faster than the previous implementation. This is because we continue sending file chunks in parallel even when switching files. I also verified the rate limiter with max_concurrent_file_chunks with max_bytes_per_sec=10mb, 40mb and 100mb. It works perfectly. You can find the result [here](https://gist.github.com/dnhatn/1aa4bd47513c746643b5076fe942ac8a).

I have responded to your comments. Please give this PR another shot. Thank you.

**Comment by dnhatn** [comment]
@elasticmachine run gradle build tests 1

File: docs/reference/modules/indices/recovery.asciidoc
Comment (ywelsch) [inline docs/reference/modules/indices/recovery.asciidoc]: Controls the number of file chunk requests *that are* sent in parallel per *peer* recovery.

File: docs/reference/modules/indices/recovery.asciidoc
Comment (ywelsch) [inline docs/reference/modules/indices/recovery.asciidoc]: Perhaps change this to: "As multiple recoveries are already running in parallel (controlled by `cluster.routing.allocation.node_concurrent_recoveries`), increasing this expert-level setting might only help in situations where peer recovery of a single shard is not reaching the total inbound and outbound peer recovery traffic as configured by `indices.recovery.max_bytes_per_sec`, but is CPU-bound instead, typically when using transport-level security or compression."

**Review by ywelsch — APPROVED** [review]

**Comment by dnhatn** [comment]
run gradle build tests 1
run gradle build tests 2

**Comment by dnhatn** [comment]
@ywelsch Thanks so much for proposing this idea and reviewing. Thanks @original-brownbear.

**Pull request merged**
